### ISSUE TITLE
Template not respected as part of `@method`

### ISSUE THREAD

**Issue by jmfederico** [vimeo/psalm#10226]
Example code:

```php
<?php

/**
 * @template-extends \Traversable<mixed>
 */
interface ExtendsTraversable extends Traversable
{
}

/**
 * @method \ExtendsTraversable<string> paginate()
 */
class Foo {
    public function bar(): void
    {
        foreach ($this->paginate() as $baz) {
            echo $baz;
        }
    }
}
```

### Expected behaviour:
- `$baz` is detected as a `string`

### Observed behaviour:
- `$baz` is detected as a `mixed`

**Comment by jmfederico** [comment]
> Hey @jmfederico, can you reproduce the issue on https://psalm.dev ?

Yes

**Comment by ging-dev** [comment]
> > Hey @jmfederico, can you reproduce the issue on https://psalm.dev ?
> 
> Yes

https://psalm.dev/r/09caf60fe7

**Comment by someniatko** [comment]
@jmfederico You have not defined your `ExtendsTraversable` interface as one having a template. Instead, it has a fixed `mixed` type which is forwarded to parent - so it is always `Traversable<mixed>`. To be able to type-parameterize `ExtendsTraversable`, you have to add `@template T` (or some other name instead of `T`) to the `ExtendsTraversable`, and instead of `@template-extends Traversable<mixed>` use `@template-extends Traversable<T>`. Now you can parameterize `ExtendsTraversable` itself.

**Comment by jmfederico** [comment]
Other tools interpret this as "expected", which is why I brought it up here.

But also, have a look at the screenshot I am attaching. It seems like the `paginate` method is recognized by Psalm as outputting strings, but the variable is then interpreted as a mixed.

<img width="691" alt="Skärmavbild 2023-09-28 kl  10 35 01" src="https://github.com/vimeo/psalm/assets/237977/dcce3876-573a-4075-a361-5bd17558f66d">

**Comment by robchett** [comment]
`$this->paginate()` is returning an `\ExtendsTraversable<string>` as per the `@method` annotation but `\ExtendsTraversable` has no template parameters so it is still a `\Traversable<mixed>`

That's why you need to use 

```php
/**
 * @template T
 * @extends \Traversable<T>
 */
interface  ExtendsTraversable extends \Traversable
{
}
```

All of this is in-line with Psalms documented usage of extended templates 

```php
<?php
/**
 * @template T1
 */
class ParentClass {}

/**
 * @template T2
 * @extends ParentClass<T2>
 */
class ChildClass extends ParentClass {}
```
